dot/parachain/availability-store/availability_store.go

```diff
@@ -604,12 +836,275 @@ func (av *AvailabilityStoreSubsystem) handleStoreAvailableData(msg StoreAvailabl
+			stateValue, err := meta.State.Value()
+			if err != nil {
+				logger.Errorf("failed to get state value: %w", err)
+			}
+			switch val := stateValue.(type) {
+			case Finalized:
+				continue // sanity check
+			case Unavailable:
+				// This is also not going to happen; the very fact that we are
+				// iterating over the candidate here indicates that `State` should
+				// be `Unfinalized`.
+				err = tx.pruneByTime.Del(append(uint32ToBytes(uint32(val.Timestamp)), candidateHash.Value[:]...))
+				if err != nil {
+					logger.Errorf("failed to delete pruning key: %w", err)
+				}
+			case Unfinalized:
+				for _, v := range val.BlockEntry {
+					if v.BlockNumber != parachaintypes.BlockNumber(blockNumber) {
+						// deleteUnfinalizedInclusion
+						key := append(uint32ToBytes(uint32(v.BlockNumber)), v.BlockHash[:]...)
+						key = append(key, candidateHash.Value[:]...)
+						err = tx.unfinalized.Del(key)
+						if err != nil {
+							logger.Errorf("failed to delete unfinalized key: %w", err)
+						}
+					}
+				}
+			}
+
+			err = meta.State.Set(Finalized{Timestamp: now})
+			if err != nil {
+				logger.Errorf("failed to set state to finalized: %w", err)
+			}
+
+			// write meta
+			err = writeCandidateMetaToDB(tx.meta, candidateHash, meta)
+			if err != nil {
+				logger.Errorf("storing metadata for candidate %v: %w", candidateHash, err)
+			}
+
+			// write pruning key
+			pruneAt := now + BETimestamp(av.pruningConfig.KeepFinalizedFor.Seconds())
+			err = av.writePruningKey(tx.pruneByTime, pruneAt, candidateHash)
+			if err != nil {
+				logger.Errorf("writing pruning key: %w", err)
+			}
+
+		} else {
+			stateValue, err := meta.State.Value()
+			if err != nil {
+				logger.Errorf("failed to get state value: %w", err)
+			}
+			switch val := stateValue.(type) {
+			case Finalized:
+				continue // sanity
+			case Unavailable:
+				continue // sanity
+			case Unfinalized:
+				retainedBlocks := []BlockEntry{}
+				for _, v := range val.BlockEntry {
+					if v.BlockNumber != parachaintypes.BlockNumber(blockNumber) {
+						retainedBlocks = append(retainedBlocks, v)
+					}
+				}
+				if len(retainedBlocks) == 0 {
+					// write pruning key
+					pruneAt := val.Timestamp + BETimestamp(av.pruningConfig.KeepUnavailableFor.Seconds())
+					err = av.writePruningKey(tx.pruneByTime, pruneAt, candidateHash)
+					if err != nil {
+						logger.Errorf("writing pruning key: %w", err)
+					}
+
+					err = meta.State.Set(Unavailable{Timestamp: val.Timestamp})
+					if err != nil {
+						logger.Errorf("failed to set state to unavailable: %w", err)
+					}
+				} else {
+					err = meta.State.Set(Unfinalized{Timestamp: val.Timestamp, BlockEntry: retainedBlocks})
+					if err != nil {
+						logger.Errorf("failed to set state to unfinalized: %w", err)
+					}
+				}
+			}
+
+			// write meta
+			err = writeCandidateMetaToDB(tx.meta, candidateHash, meta)
+			if err != nil {
+				logger.Errorf("failed to put meta: %w", err)
+			}
+		}
+	}
+}
```

Comment:
There are a lot of errors not being handled here.  I understand that we want to continue processing candidates even if there are some errors, but are there no errors that should eventually cause this subsystem to shut down? I would assume if we're unable to write to the DB that could be a systemic error.  Just returning the error and panicking/logging on the caller side for now will make things easier when we implement graceful shutdown.

Reply:
Added panics where systemic errors are reached.